### PR TITLE
Merge in outstanding staging commits

### PR DESCRIPTION
**Pull request by johnoliver**
# Checklist

- [ ] You added tests to cover the change
- [ ] `mvn clean install` build and test completes
- [ ] You changed or added to the documentation

### CODE REVIEW

**Comment by nickebbitt** [comment]
looking forward to this, maybe due to the PR size it's worth a call so you can talk me (anyone else interested) through the changes? if so, give me a shout when it's ready and we'll find a good time

File: adoptopenjdk-api-v3-frontend/src/main/kotlin/net/adoptopenjdk/api/v3/routes/packages/BinaryResource.kt [hunk 1]
```diff
@@ -21,11 +11,9 @@ import org.eclipse.microprofile.openapi.annotations.responses.APIResponses
...
 import org.jboss.resteasy.annotations.jaxrs.PathParam
 import org.jboss.resteasy.annotations.jaxrs.QueryParam
-import javax.ws.rs.GET
-import javax.ws.rs.HEAD
-import javax.ws.rs.HeaderParam
-import javax.ws.rs.Path
-import javax.ws.rs.Produces
+import javax.enterprise.context.ApplicationScoped
+import javax.inject.Inject
+import javax.ws.rs.*
```
Comment (karianna) [inline adoptopenjdk-api-v3-frontend/src/main/kotlin/net/adoptopenjdk/api/v3/routes/packages/BinaryResource.kt:16]: explicit imports

File: adoptopenjdk-api-v3-frontend/src/main/kotlin/net/adoptopenjdk/api/v3/routes/packages/InstallerResource.kt [hunk 2]
```diff
@@ -1,17 +1,7 @@
...
-import net.adoptopenjdk.api.v3.models.HeapSize
-import net.adoptopenjdk.api.v3.models.ImageType
-import net.adoptopenjdk.api.v3.models.Installer
-import net.adoptopenjdk.api.v3.models.JvmImpl
-import net.adoptopenjdk.api.v3.models.OperatingSystem
-import net.adoptopenjdk.api.v3.models.Project
-import net.adoptopenjdk.api.v3.models.Release
-import net.adoptopenjdk.api.v3.models.ReleaseType
-import net.adoptopenjdk.api.v3.models.Vendor
+import net.adoptopenjdk.api.v3.models.*
```
Comment (karianna) [inline adoptopenjdk-api-v3-frontend/src/main/kotlin/net/adoptopenjdk/api/v3/routes/packages/InstallerResource.kt:4]: imports

File: adoptopenjdk-api-v3-frontend/src/main/kotlin/net/adoptopenjdk/api/v3/routes/packages/PackageEndpoint.kt [hunk 3]
```diff
@@ -7,22 +7,13 @@ import net.adoptopenjdk.api.v3.dataSources.SortOrder
...
-import net.adoptopenjdk.api.v3.models.Binary
-import net.adoptopenjdk.api.v3.models.HeapSize
-import net.adoptopenjdk.api.v3.models.ImageType
-import net.adoptopenjdk.api.v3.models.JvmImpl
-import net.adoptopenjdk.api.v3.models.OperatingSystem
-import net.adoptopenjdk.api.v3.models.Project
-import net.adoptopenjdk.api.v3.models.Release
-import net.adoptopenjdk.api.v3.models.ReleaseType
-import net.adoptopenjdk.api.v3.models.Vendor
+import net.adoptopenjdk.api.v3.models.*
```
Comment (karianna) [inline adoptopenjdk-api-v3-frontend/src/main/kotlin/net/adoptopenjdk/api/v3/routes/packages/PackageEndpoint.kt:10]: imports

File: adoptopenjdk-api-v3-frontend/src/main/kotlin/net/adoptopenjdk/api/v3/routes/stats/DownloadStatsResource.kt [hunk 4]
```diff
@@ -5,13 +5,8 @@ import kotlinx.coroutines.launch
...
 import net.adoptopenjdk.api.v3.TimeSource
 import net.adoptopenjdk.api.v3.dataSources.APIDataStore
-import net.adoptopenjdk.api.v3.dataSources.ApiPersistenceFactory
 import net.adoptopenjdk.api.v3.dataSources.models.FeatureRelease
-import net.adoptopenjdk.api.v3.models.JvmImpl
-import net.adoptopenjdk.api.v3.models.Release
-import net.adoptopenjdk.api.v3.models.ReleaseType
-import net.adoptopenjdk.api.v3.models.StatsSource
-import net.adoptopenjdk.api.v3.models.Vendor
+import net.adoptopenjdk.api.v3.models.*
```
Comment (karianna) [inline adoptopenjdk-api-v3-frontend/src/main/kotlin/net/adoptopenjdk/api/v3/routes/stats/DownloadStatsResource.kt:9]: imports

**Review by karianna — APPROVED** [review]
Nitpick on the explicit imports, rest looks good.

**Review by nickebbitt — APPROVED** [review]
Looks good, I'll look to help out migrating other areas of the code base to this approach too

**Comment by karianna** [comment]
@johnoliver Can be taken out of draft?

**Comment by johnoliver** [comment]
@nickebbitt @karianna this PR now brings in all outstanding changes from staging:
- Adds DI implementation
- Logging to Application Insights
- Add `/release_name/{vendor}/{release_name}` path
- Tidy poms
- Bump lots of dependencies
  - Kotlin -> 1.4.20
  - Quarkus -> 1.10.0
- Remove dependency on live data for testing

**Comment by nickebbitt** [comment]
thanks, I'll take a look through today 👍

File: adoptopenjdk-api-v3-models/src/main/kotlin/net/adoptopenjdk/api/v3/dataSources/models/GitHubId.kt [hunk 5]
```diff
@@ -18,7 +18,7 @@ class GithubId {
         if (this === other) return true
         if (javaClass != other?.javaClass) return false
 
-        other as GithubId
+        other as GitHubId
 
         if (githubId != other.githubId) return false
```
Comment (nickebbitt) [inline adoptopenjdk-api-v3-models/src/main/kotlin/net/adoptopenjdk/api/v3/dataSources/models/GitHubId.kt:23]: really nit-picky: `gitHubId`

File: adoptopenjdk-api-v3-models/src/main/kotlin/net/adoptopenjdk/api/v3/dataSources/models/Releases.kt [hunk 6]
```diff
@@ -62,18 +62,18 @@ class Releases {
...
         return nodes
-            .any { it.key.startsWith(githubId.githubId) }
+            .any { it.key.startsWith(gitHubId.githubId) }
     }
 
-    fun hasReleaseBeenUpdated(githubId: GithubId, updatedAt: ZonedDateTime): Boolean {
+    fun hasReleaseBeenUpdated(gitHubId: GitHubId, updatedAt: ZonedDateTime): Boolean {
         return nodes
-            .filter { it.key.startsWith(githubId.githubId) }
+            .filter { it.key.startsWith(gitHubId.githubId) }
```
Comment (nickebbitt) [inline adoptopenjdk-api-v3-models/src/main/kotlin/net/adoptopenjdk/api/v3/dataSources/models/Releases.kt:76]: Not for this review really, but `gitHubId.githubId` is a bit weird.

File: adoptopenjdk-api-v3-updater/src/main/kotlin/net/adoptopenjdk/api/v3/V3Updater.kt [hunk 7]
```diff
@@ -130,6 +149,10 @@ class V3Updater {
             }
         } catch (e: Exception) {
             LOGGER.error("Failed to perform full update", e)
+        } catch (e: Throwable) {
```
Comment (nickebbitt) [inline adoptopenjdk-api-v3-updater/src/main/kotlin/net/adoptopenjdk/api/v3/V3Updater.kt:152]: Just wondering what the distinction it between this catch and the one just above. Why do we swallow when it's `Exception` but re-throw when it's `Throwable`? 

Might be worth a comment or consolidating if we actually want to handle them in the same way.

File: adoptopenjdk-api-v3-updater/src/test/kotlin/net/adoptopenjdk/api/TestResourceDouble.kt [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+import net.adoptopenjdk.api.v3.models.Project
+import net.adoptopenjdk.api.v3.models.Release
+import net.adoptopenjdk.api.v3.models.ReleaseType
+import net.adoptopenjdk.api.v3.models.SourcePackage
+import net.adoptopenjdk.api.v3.models.Vendor
+import net.adoptopenjdk.api.v3.models.VersionData
+import java.time.ZonedDateTime
+import java.util.Random
+
+object TestResourceDouble {
```
Comment (nickebbitt) [inline adoptopenjdk-api-v3-updater/src/test/kotlin/net/adoptopenjdk/api/TestResourceDouble.kt:24]: Is this really a test double, or more just a collection of test data?

**Review by nickebbitt — APPROVED** [review]
lgtm - I've left a few comments but there's nothing in there that I'd not merge for

File: adoptopenjdk-api-v3-frontend/src/main/kotlin/net/adoptopenjdk/api/v3/routes/info/PlatformsResource.kt [hunk 9]
```diff
@@ -21,6 +21,6 @@ class PlatformsResource {
     // Hide this path as it is only used internally by the website
     @Operation(summary = "Returns information about available platforms", hidden = true)
     fun get(): Platforms {
-        return APIDataStore.platforms
+        return VariantStore.platforms
```
[the remark this answers is not in the document]
Comment (johnoliver) [reply adoptopenjdk-api-v3-frontend/src/main/kotlin/net/adoptopenjdk/api/v3/routes/info/PlatformsResource.kt:24]: It certainly would, I kind of wanted to just get the overall structure of the DI agreed to before we went through and applied it everywhere.

[the remark this answers is not in the document]
Comment (nickebbitt) [reply adoptopenjdk-api-v3-frontend/src/main/kotlin/net/adoptopenjdk/api/v3/routes/info/PlatformsResource.kt:24]: cool, yeah agreed, we should get this merged and then can migrate as we touch various parts of the code base going forward

File: adoptopenjdk-api-v3-updater/src/main/kotlin/net/adoptopenjdk/api/v3/V3Updater.kt (hunk 7 above)
Comment (johnoliver) [reply adoptopenjdk-api-v3-updater/src/main/kotlin/net/adoptopenjdk/api/v3/V3Updater.kt:152]: Throwable can include unrecoverable errors that we want to log but then allow the VM to fall over and get restarted such as OutOfMemoryError

File: adoptopenjdk-api-v3-frontend/src/test/kotlin/net/adoptopenjdk/api/DbExtension.kt [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+import org.junit.jupiter.api.extension.BeforeAllCallback
+import org.junit.jupiter.api.extension.ExtensionContext
+
+class DbExtension : BeforeAllCallback, AfterAllCallback {
+    override fun beforeAll(context: ExtensionContext?) {
+        System.setProperty(ENABLE_PERIODIC_UPDATES, "false")
+    }
+
+    override fun afterAll(p0: ExtensionContext?) {
+        System.gc() // Don't ask, but also don't remove me, breaks deadlock that hangs vm after all tests are completed
```
[the remark this answers is not in the document]
Comment (johnoliver) [reply adoptopenjdk-api-v3-frontend/src/test/kotlin/net/adoptopenjdk/api/DbExtension.kt:14]: There is some bug where the vm does not exit due to threads not exiting, forcing a System.gc clears this

**Comment by karianna** [comment]
@johnoliver Can this be taken out of draft?

**Comment by nickebbitt** [comment]
if this is good to go, would you like me to do the honours and merge?

**Comment by johnoliver** [comment]
yup go for it

**Comment by johnoliver** [comment]
make sure to squash merge, dont want all those commits

**Pull request merged**